[docs] LibraryEvolution: Non-controversial updates for Swift 4.2

jrose-apple opened the pull request:
- Standardize on "inlinable" instead of "inlineable", per [SE-0193][].
- Standardize on "frozen enums" instead of "closed enums", even though [SE-0192][] wasn't quite conclusive on this.
- Typealiases are treated as versioned entities for the purpose of verifying availability.
- Per [SE-0193][], inlinable functions do not have always-emit-into-client behavior.
- Rename the fragility attribute for properties and subscripts to `@inlinableAccess` for now, since `@inlinable` has been implemented for functions now. (We may still end up choosing to use the same name for these declarations too.)
- Private and fileprivate entities cannot be `@inlinable` today.
- We implemented the always-emit-into-client change for default arguments.
- Disallow adding new associated types to protocols for now.
- ~Disallow adding setters to 'open' properties for now; we have no good way to backwards-deploy such a change.~
- Mention that precedence groups exist.
- Mention that `@usableFromInline` exists.
- `@inlinable` can be backdated, even though all the other fragility attributes cannot.
- Tweak TODO around `@NSManaged`.
- Tweak list of planned proposals.
- A struct initializer can also assign to `self` instead of delegating with `self.init`. (See [SE-0189][] for more details.)

This document still deserves some more drastic restructuring to separate the near-term features from the long-term features.

[SE-0189]: https://github.com/apple/swift-evolution/blob/master/proposals/0189-restrict-cross-module-struct-initializers.md
[SE-0192]: https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md
[SE-0193]: https://github.com/apple/swift-evolution/blob/master/proposals/0193-cross-module-inlining-and-specialization.md

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test Linux

slavapestov commented on docs/LibraryEvolution.rst:
```diff
@@ -409,40 +393,29 @@ polar representation::
-Default argument expressions are implemented as ``@inlineable`` functions and
-thus are subject to the same restrictions as inlineable functions. A default
+Default argument expressions are implemented as ``@inlinable`` functions and
+thus are subject to the same restrictions as inlinable functions. A default
```
Only if the function is public, `@usableFromInline` or `@inlinable` -- otherwise there is no restriction.

And when it's public, the restriction is stricter -- it can only reference _public_ declarations. Referencing `@usableFromInline` is not allowed.

jrose-apple replied:
Whoops, yep. Thanks.

slavapestov commented on docs/LibraryEvolution.rst:
```diff
@@ -513,29 +487,29 @@ clients to access them more efficiently. This restricts changes a fair amount:
+.. admonition:: TODO
+
+    ``@inlinableAccess`` isn't implemented yet, but for computed properties we
+    already allow putting ``@inlinable`` on the accessors individually. That
```
Currently implemented:

- Computed properties can be `@inlinable` (or individual accessors, as you say)
- Stored properties can be `@_fixed_layout` (which mostly does what you describe for `@inlinableAccess`)

jrose-apple replied:
Yeah. I'm not clear on how these compose, but maybe they don't need to. Stored properties with observers, maybe?

slavapestov commented on docs/LibraryEvolution.rst:
```diff
@@ -513,29 +487,29 @@ clients to access them more efficiently. This restricts changes a fair amount:
 Note that if a constant's initial value expression has any observable side
 effects, including the allocation of class instances, it must not be treated
-as inlineable. A constant must always behave as if it is initialized exactly
+as inlinable. A constant must always behave as if it is initialized exactly
```
Why not?

jrose-apple replied:
Global / static initial value expressions are a replacement for `dispatch_once` and `std::call_once`. They had better not happen more than once!

slavapestov commented on docs/LibraryEvolution.rst:
```diff
@@ -607,9 +581,9 @@ stored subscripts. This means that the following changes are permitted:
-the body of an accessor a `binary-compatible source-breaking change`. Any
-inlineable accessors must follow the rules for `inlineable functions`_, as
+Like properties, subscripts can be marked ``@inlinableAccess``, which makes
+changing the body of an accessor a `binary-compatible source-breaking change`.
```
Why is it source breaking? What if I just optimize an algorithm inside my accessor body?

jrose-apple replied:
It's not source-incompatible, sure. Maybe we need a different category name for "binary-compatible, but still has lasting effects across versions".

slavapestov commented on docs/LibraryEvolution.rst:
```diff
@@ -781,9 +756,9 @@ Initializers
+``@inlinable``, with the same restrictions. Unlike struct initializers, enum
 initializers do not always need to delegate to another initializer, even if
-they are inlineable or declared in a separate module.
+they are inlinable or declared in a separate module.
```
Enum initializers either delegate to another initializer or assign to `self`. Both are effectively equivalent.

jrose-apple replied:
Yeah. I didn't feel like that was interesting to call out here.

slavapestov commented on docs/LibraryEvolution.rst:
```diff
@@ -700,8 +674,8 @@ still be modified in limited ways:
-An initializer of a fixed-contents struct may be declared ``@inlineable`` even
-if it does not delegate to another initializer, as long as the ``@inlineable``
+An initializer of a fixed-contents struct may be declared ``@inlinable`` even
+if it does not delegate to another initializer, as long as the ``@inlinable``
```
If the struct is not fixed-contents the initializer doesn't have to delegate to another initializer -- it can also assign to `self`.

jrose-apple replied:
Ah, right.

slavapestov commented on docs/LibraryEvolution.rst:
```diff
@@ -861,25 +835,23 @@ All other changes to the protocol itself are forbidden, including:
-- Adding or removing constraints from an associated type.
-- Removing a default from an associated type.
+- Adding or removing constraints from an associated type, including inherited
+  associated types.
```
Maybe call out adding associated types in this list?

slavapestov commented on docs/LibraryEvolution.rst:
```diff
@@ -1016,19 +994,16 @@ little. They allow the following changes:
-    struct/enum methods and top-level functions: a method on a class would be
-    part of the library's module interface even if it had been inlineable since
-    its introduction.
+    marked ``@inlinable``, permitting inlining based on speculative
```
This restriction is not implemented today and the standard library has non-final inlinable class methods. Can we remove it?

(Note that `dynamic @inlinable` is still an error, since there's no way to statically call such a method.)

jrose-apple committed:
[docs] LibraryEvolution: Non-controversial updates for Swift 4.2

- Standardize on "inlinable" instead of "inlineable", per SE-0193.
- Standardize on "frozen enums" instead of "closed enums", even though
  SE-0192 wasn't quite conclusive on this.
- Typealiases are treated as versioned entities for the purpose of
  verifying availability.
- Per SE-0193, inlinable functions do not have always-emit-into-client
  behavior.
- Rename the fragility attribute for properties and subscripts to
  '@inlinableAccess' for now, since '@inlinable' has been implemented
  for functions now. (We may still end up choosing to use the same
  name for these declarations too.)
- Private and fileprivate entities cannot be '@inlinable' today.
- We implemented the always-emit-into-client change for default
  arguments. They also have even stricter restrictions than most
  inlinable functions.
- A struct initializer can also assign to 'self' instead of delegating
  with `self.init`.
- Disallow adding new associated types to protocols for now.
- Mention that precedence groups exist.
- Mention that '@usableFromInline' exists.
- '@inlinable' can be backdated, even though all the other fragility
  attributes cannot.
- Tweak TODO around '@NSManaged'.
- Tweak list of planned proposals.

This document still deserves some more drastic restructuring to
separate the near-term features from the long-term features.

slavapestov reviewed (commented):
Looks good!
